[ManagedMSE] JS ManagedSourceBuffer object doesn't present any of its attributes

jyavenard opened the pull request:
#### 202d7e72c2c6566cff1ed27fa4d010820598c51d
<pre>
[ManagedMSE] JS ManagedSourceBuffer object doesn&apos;t present any of its attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=257037">https://bugs.webkit.org/show_bug.cgi?id=257037</a>
rdar://109569570

Reviewed by Chris Dumez.

Add missing custom toJS for MediaSource and SourceBuffer.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html:
* Source/WebCore/Modules/mediasource/MediaSource.idl:
* Source/WebCore/Modules/mediasource/SourceBuffer.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSMediaSourceCustom.cpp: Added.
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
* Source/WebCore/bindings/js/JSSourceBufferCustom.cpp: Added.
(WebCore::toJSNewlyCreated):
(WebCore::toJS):

Canonical link: <a href="https://commits.webkit.org/264284@main">https://commits.webkit.org/264284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2252eac6adb781b627d240bc21e78e0de0d3b43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7253 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7503 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8874 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7461 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10356 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7380 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8074 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6670 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8982 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5417 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6597 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14330 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9588 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7185 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5868 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6542 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6507 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1716 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10743 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6923 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/media/media-source/media-managedmse-bufferedchange.html:
```diff
@@ -37,6 +37,10 @@
             run('sourceBuffer.appendBuffer(loader.initSegment())');
             await waitFor(sourceBuffer, 'update');
 
+            sourceBuffer.onbufferedchange = () => {
```
Testing isn't great. I'd add something like:
```
shouldBe("sourceBuffer.__proto__", "ManagedSourceBuffer.prototype");
shouldBeTrue("sourceBuffer.hasOwnProperty('onbufferedchange')");
```

Or whatever the equivalent is in media tests.

A test that doesn't FAIL when the property is missing is not a good test.

jyavenard replied:
it would have failed if the property was missing.
That's how I found the bug to start.

If the sourceBuffer object is actually a `SourceBuffer` rather than a `ManagedSourceBuffer` then the `onbufferedchange` property doesn't exist to start, and so setting it does nothing.

cdumez replied:
It wouldn't have failed, it would have given *different* output which could have easily been mistakenly rebaselined. It is bad practice. Please make it so that the test output contains a FAIL line if the property is missing.

cdumez commented on Source/WebCore/bindings/js/JSMediaSourceCustom.cpp:
```diff
@@ -0,0 +1,50 @@
+
+#include "config.h"
+
+#if ENABLE(MANAGED_MEDIA_SOURCE)
```
This doesn't seem right.

jyavenard replied:
oh right...

cdumez commented on Source/WebCore/bindings/js/JSMediaSourceCustom.cpp:
```diff
@@ -0,0 +1,50 @@
+
+JSValue toJSNewlyCreated(JSGlobalObject*, JSDOMGlobalObject* globalObject, Ref<MediaSource>&& source)
+{
+    if (source->isManaged())
```
Seems only this line and the next should be guarded with `ENABLE(MANAGED_MEDIA_SOURCE)`

cdumez commented on Source/WebCore/bindings/js/JSSourceBufferCustom.cpp:
```diff
@@ -0,0 +1,50 @@
+
+JSValue toJSNewlyCreated(JSGlobalObject*, JSDOMGlobalObject* globalObject, Ref<SourceBuffer>&& buffer)
+{
+    if (buffer->isManaged())
```
ditto.

cdumez commented on Source/WebCore/bindings/js/JSSourceBufferCustom.cpp:
```diff
@@ -0,0 +1,54 @@
+#if ENABLE(MANAGED_MEDIA_SOURCE)
+#include "JSManagedSourceBuffer.h"
+#endif
+#include "JSSourceBuffer.h"
```
First come the non guarded includes, then the guarded ones with a blank line in between, so it should look like:
```
#include "JSSourceBuffer.h"

#if ENABLE(MANAGED_MEDIA_SOURCE)
#include "JSManagedSourceBuffer.h"
#endif
```

cdumez commented on LayoutTests/media/media-source/media-managedmse-bufferedchange.html:
```diff
@@ -35,6 +35,9 @@
 
             run('sourceBuffer = source.addSourceBuffer(loader.type())');
+            testExpected('sourceBuffer.__proto__', ManagedSourceBuffer.prototype);
+            testExpected('sourceBuffer.onbufferedchange !== undefined', true);
```
My comment was to add these checks in addition to your new event listener. Checking that the event listener gets called seemed useful too.

jyavenard replied:
indeed ; but right now there's another bug that makes it behave differently if the GPU process is turned on. I will fix it soon

cdumez replied:
Oh, right. Fine.

cdumez reviewed (approved):
LGTM % nits.

jyavenard committed:
[ManagedMSE] JS ManagedSourceBuffer object doesn't present any of its attributes
https://bugs.webkit.org/show_bug.cgi?id=257037
rdar://109569570

Reviewed by Chris Dumez.

Add missing custom toJS for MediaSource and SourceBuffer.

* LayoutTests/media/media-source/media-managedmse-bufferedchange-expected.txt:
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html:
* Source/WebCore/Modules/mediasource/MediaSource.idl:
* Source/WebCore/Modules/mediasource/SourceBuffer.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSMediaSourceCustom.cpp: Added.
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
* Source/WebCore/bindings/js/JSSourceBufferCustom.cpp: Added.
(WebCore::toJSNewlyCreated):
(WebCore::toJS):

Canonical link: https://commits.webkit.org/264284@main
